aten/src/ATen/cuda/CUDABlas.cpp

```
@@ -1067,8 +1226,161 @@ void gemm_internal_cublas<at::BFloat16>(CUDABLAS_GEMM_ARGTYPES(at::BFloat16)) {
   TORCH_CUDABLAS_CHECK(cublasSetMathMode(handle, CUBLAS_DEFAULT_MATH));
 }
 
-template <typename Dtype>
-inline void gemm_internal_cublaslt(CUDABLAS_GEMM_ARGTYPES(Dtype)) {
+template <>
+void gemm_internal_cublas<at::Half, float>(CUDABLAS_GEMM_ARGTYPES_AND_C_DTYPE(at::Half, float)) {
+  // See Note [Writing Nondeterministic Operations]
+  globalContext().alertCuBLASConfigNotDeterministic();
+  cublasHandle_t handle = at::cuda::getCurrentCUDABlasHandle();
+  cublasOperation_t opa = _cublasOpFromChar(transa);
+  cublasOperation_t opb = _cublasOpFromChar(transb);
+  float falpha = alpha;
+  float fbeta = beta;
+#ifndef USE_ROCM
+  at::Half halpha;
+  at::Half hbeta;
+  auto compute_type = CUDA_R_32F;
+#endif
+  void * alpha_ptr = &falpha;
+  void * beta_ptr = &fbeta;
+  _cublasAdjustLdLevel3(transa, transb, m, n, k, &lda, &ldb, &ldc);
+  GEMM_CHECK_ARGVALUES(at::Half);
+#ifdef USE_ROCM
+  int flag = 0;
+#if USE_GEMM_FLAGS_FP16_ALT_IMPL
+  flag = at::ROCmBackwardPassGuard::is_backward_pass() ? rocblas_gemm_flags_fp16_alt_impl : 0;
+#endif
+  TORCH_CUDABLAS_CHECK(rocBLASStatusToHIPStatus(rocblas_gemm_ex(
+      (rocblas_handle)handle,
+      hipOperationToRocOperation(opa),
+      hipOperationToRocOperation(opb),
+      m,
+      n,
+      k,
+      alpha_ptr,
+      a,
+      rocblas_datatype_f16_r,
+      lda,
+      b,
+      rocblas_datatype_f16_r,
+      ldb,
+      beta_ptr,
+      c,
+      rocblas_datatype_f16_r,
+      ldc,
+      c,
+      rocblas_datatype_f16_r,
+      ldc,
+      rocblas_datatype_f32_r,
+      rocblas_gemm_algo_standard,
+      0,
+      flag)));
+#else
+  cudaDeviceProp* prop = at::cuda::getCurrentDeviceProperties();
+  if (prop->major >= 7 && at::globalContext().allowFP16AccumulationCuBLAS()) {
```

Same situation as above, does it make sense to add output `dtype` as a param when accumulation was done in `half`?

---

I think if cublas itself allows this combination, there's no reason to throw an error for it. If on the other hand cublas doesn't support it then throwing an error seems more reasonable than coming up with workarounds.

---

@ngimel @eqy will get this combination to throw an error as fp16 accum with fp32 output isn't actually supported in cublas based on docs: https://docs.nvidia.com/cuda/cublas/#cublasgemmstridedbatchedex

---

I'm not sure I follow. As @ngimel pointed out, the unfortunately named `allowFP16Accumulation` and `allowBF16Accumulation` flags actually control the math mode, and if they are `true`, we would set e.g., 
`cublasSetMathMode(handle, CUBLAS_DEFAULT_MATH | CUBLAS_MATH_DISALLOW_REDUCED_PRECISION_REDUCTION)` https://docs.nvidia.com/cuda/cublas/#cublassetmathmode. Which I assume since it allows both---would be surprising if it raises a unsupported error.

---

`allowFP16AccumulationCuBLAS` will set compute type to CUBLAS_COMPUTE_16F, and according to the docs, only fp16 Ctype is compatible with that, so we should throw an error if user-requested ctype  is fp32. `allowFP16ReductionCuBLAS` sets MathMode and should be compatible with different ctypes. So I think @PaulZhang12 is on the right track